ext/ox/obj_load.c

```diff
@@ -739,15 +739,38 @@ end_element(PInfo pi, const char *ename) {
 		helper_stack_push(&pi->helpers, h->var, h->obj, KeyCode);
 		break;
 	    case RangeCode:
-		if (ox_beg_id == h->var) {
-		    RSTRUCT_SET(ph->obj, 0, h->obj);
-		} else if (ox_end_id == h->var) {
-		    RSTRUCT_SET(ph->obj, 1, h->obj);
-		} else if (ox_excl_id == h->var) {
-		    RSTRUCT_SET(ph->obj, 2, h->obj);
-		} else {
-		    set_error(&pi->err, "Invalid range attribute", pi->str, pi->s);
-		    return;
+		{
+		    VALUE begin, end;
+		    int excl;
+
```

Comment:
I hope you don't mind me being a little picky about code format.

Please use tabes to line up the arguments.

More important, all three variable must be initialized. I'd suggest:

``` ruby
	VALUE	begin = Qnil;
	VALUE	end = Qnil;
	int	excl = 0;
```

But, I'm sorry to say there is one major flaw. Let's back up a little. The parser puts values on a stack. The first call will be will be at line 622. That currently creates a range of `0..0`. Since in Ruby 3.0 the Range can not be modified another type must be used to collect the begin. end, and excl. Since the type must be a VALUE then something like a Ruby **struct**, array, or hash would be a good choice for collecting the values.

In this code each condition would set either the begin, end, or excl in the temporary object.

One final change. You might have noticed that the `end_element()` call doe not modify the popped element from the stack because all the objects are ready to go at all times. This will no longer be the case with a Range object so a new Range object will need to be created so the fully formed Range can but added to the parent.